libp2p/protocols/perf/server.nim

```
@@ -24,7 +24,7 @@ type Perf* = ref object of LPProtocol
 
 proc new*(T: typedesc[Perf]): T {.public.} =
   var p = T()
-  proc handle(conn: Connection, proto: string) {.async.} =
+  proc handle(conn: Connection, proto: string) {.async: (raises: []).} =
```

I'm not convinced having the handlers swallow cancellations is a good idea - when disconnecting a peer, the owner of the future should cancel any handlers and that should propagate all the way. What's the reasoning here?

---

owner of `handlers` will not know how to handle exceptions from `handlers` (should they raise). and it's up to `handlers` to handle them as best as possible. this was the reason of not propagating exceptions.

cancellation: owner of `handle` futures can cancel them, and `handler` will gracefully handle cancellation of it's logic inside. futures will finish (as early as possible) giving control to owner.

---

it's fine not to raise other errors, but propagating cancellations is _probably_ the better thing to do in this case - when you are the owner and perform a cancellation, the exception gets handled automatically by chronos at the right abstraction layer.

The rare exception in which we don't propagate cancellations is for "top-level" tasks that are meant to be spawned (a bit like the main procedure of a thread).

---

i see handlers as "top-level" tasks as well... 

okay, i realize point you are tiring to make. let me investigate bit into this matter.

---

yeah, handlers tend to be somewhere in-between top-level and not top-level - I'm not too fond of the handlers design in general because they "invert" the async abstraction a bit.. but you can imagine a web server where each request is a "handler", in which case you have something like
```
proc myserver {.async ....}
  var handlers: seq[Handler]
  try:
    while true: 
      handlers.add makehandler(await nextrequest())
  except CancelledError as e:
    await nocancel allFutures(handlers.mapIt(it.cancelAndWait())
    raise e
```

here, the handlers are not _really_ top-level the way you imagine them - the server is and you might have multiple servers which now become top-level.

It's not _wrong_ per se to have the handler catch cancellation and report "successful completion" to its owner, but it's also the more "unusual" choice, and this is maybe the key reason we let the cancellederror propagate - it makes for a recogniseable pattern.

Notably, in the above case we de facto ignore the difference between completed and cancelled handlers because we never actually _read_ the handler outcome - we just await long enough to know that there was _some_ form of completion.

---

> It's not wrong per se to have the handler catch cancellation and report "successful completion" to its owner, 

yes :+1:

> it makes for a recogniseable pattern

yes, handling cancellation the same way across code-base simplifies making decision what to do. if in some places you propagate, but in another you dont then it's harder for devs to make correct decision. in any case it is required to have firm principles for handling this.

> it's also the more "unusual" choice

it only meters if caller cares about outcome to make next decision. if next decision doesn't depend, then handler (in this case) can swallow cancelled exception and it will not have any impact (as we agree based on first statement here).

---

anyway i will proceed with raising `CancelledError`, everywhere. real top level handlers can always handle this exception and _not do anything_ or just log it.